File: new_frontend/src/services/categoryService.ts
import apiClient from "./apiClient";

export interface CategoryCreate {
  name: string;
  description?: string;
}

export interface CategoryUpdate {
  name?: string;
  description?: string;
}

export interface CategoryResponse {
  id: string;
  name: string;
  description: string | null;
  created_at: string;
}

export const categoryService = {
  getCategories: () =>
    apiClient.get<CategoryResponse[]>("/categories").then((r) => r.data),

  createCategory: (data: CategoryCreate) =>
    apiClient.post<CategoryResponse>("/categories", data).then((r) => r.data),

  updateCategory: (id: string, data: CategoryUpdate) =>
    apiClient.put<CategoryResponse>(`/categories/${id}`, data).then((r) => r.data),

  deleteCategory: (id: string) =>
    apiClient.delete<CategoryResponse>(`/categories/${id}`).then((r) => r.data),
};

File: backend/app/services/category_service.py
from sqlalchemy.orm import Session
from fastapi import HTTPException
from app.models.product import Category
from app.schemas.category import CategoryCreate, CategoryUpdate


def get_categories(db: Session):
    return db.query(Category).all()


def get_category(db: Session, category_id):
    cat = db.query(Category).filter(Category.id == category_id).first()
    if not cat:
        raise HTTPException(status_code=404, detail="Category not found")
    return cat


def create_category(db: Session, data: CategoryCreate):
    existing = db.query(Category).filter(Category.name == data.name).first()
    if existing:
        raise HTTPException(status_code=400, detail="Category already exists")
    cat = Category(**data.model_dump())
    db.add(cat)
    db.commit()
    db.refresh(cat)
    return cat


def update_category(db: Session, category_id, data: CategoryUpdate):
    cat = get_category(db, category_id)
    for key, val in data.model_dump(exclude_unset=True).items():
        setattr(cat, key, val)
    db.commit()
    db.refresh(cat)
    return cat


def delete_category(db: Session, category_id):
    cat = get_category(db, category_id)
    db.delete(cat)
    db.commit()
    return cat
